### DISCUSSION TITLE
USB Composite Device with MSC and CDC

### DISCUSSION THREAD

**Discussion by arnebdr** [zephyrproject-rtos/zephyr#70553] [Q&A] [answered]
Hi,
I want to build a USB Device with Mass Storage and CDC virtual com port. I'm using a STM32U585 on the B-U585I-IOT02A Eval Board (Zephyr Board b_u585i_iot02a).

The Mass Storage example under subsys/usb/mass is working fine. But when I try to add CDC class, windows can't enumerate the device correctly.

To build the MSC example for my board I've added an device tree overlay for the board defining the flash device:

> / {
> 	msc_disk0 {
> 		status="okay";
> 		compatible = "zephyr,flash-disk";
> 		partition = <&storage_partition>;
> 		disk-name = "NAND";
> 		/* cache-size == page erase size */
> 		cache-size = <8192>;
> 	};
> };

And then build the sample with "west build -p always -b b_u585i_iot02a samples/subsys/usb/mass -DCONFIG_APP_MSC_STORAGE_FLASH_FATFS=y"

The device is correctly recognized by the windows host and the flash drive is working fine.

To add the CDC Class I add this to the device tree overlay:

> &zephyr_udc0 {
> 	cdc_acm_uart0 {
> 		compatible = "zephyr,cdc-acm-uart";
> 	};
> };

After rebuilding and flashing, I can see that a CDC and Mass Storage Device is Plugged in, but after 30 s it removes automatically.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/164335814/7f16f891-92c3-4f51-81e0-c10b62dd5038)

The MSC Drive is not mounted in Windows and the COM Port can't be opened.
USB Tree View is reporting CM_PROB_FAILED_POST_START
![image](https://github.com/zephyrproject-rtos/zephyr/assets/164335814/fbcfeb3a-6fc9-4568-b73d-f52d07ee8aae)

The Zephyr console output does not indicate any Errors:
> [00:00:00.140,000] <inf> flash_stm32_ospi: OSPI flash config is OPI / DTR
> [00:00:00.140,000] <inf> flash_stm32_ospi: Read SFDP from octoFlash
> [00:00:00.140,000] <inf> flash_stm32_ospi: Read SFDP from octoFlash
> [00:00:00.140,000] <inf> flash_stm32_ospi: NOR octo-flash at 0x70000000 (0x4000000 bytes)
> *** Booting Zephyr OS build v3.6.0-935-g42c8b7fb00a8 ***
> [00:00:00.140,000] <inf> flashdisk: Initialize device NAND
> [00:00:00.140,000] <inf> flashdisk: offset f0000, sector size 512, page size 8192, volume size 65536
> Area 0 at 0xf0000 on flash-controller@40022000 for 65536 bytes
> [00:00:00.140,000] <inf> flashdisk: Initialize device NAND
> [00:00:00.140,000] <inf> flashdisk: offset f0000, sector size 512, page size 8192, volume size 65536
> Mount /NAND:: 0
> /NAND:: bsize = 512 ; frsize = 512 ; blocks = 80 ; bfree = 80
> /NAND: opendir: 0
> End of files
> [00:00:00.191,000] <inf> main: The device is put in USB mass storage mode.
> 
> [00:00:00.198,000] <inf> usb_cdc_acm: Device suspended
> [00:00:00.514,000] <inf> usb_cdc_acm: Device configured

Can anybody give me a hint how to get this running?

**Comment by jackchen8722** [comment]
did you fix it？

**Reply by arnebdr** [reply, marked as answer]
Hi,
yes with the new "USB Next" Device Stack this is working very well. The problems occured with the old USB Stack.
There is a "usbd_next_prj.conf" for this in the USB Mass sample.

**Reply by peterhong2016** [reply]
do we need to use usbd_msg_cb_t to configure the usb endpoint for cdc and mass device?

**Reply by jackchen8722** [reply]
Hi, i used the "USB Next" Device Stack, MSC work well, but the CDC only can receive ,can't send any data succeed...,if i config it one by one both can work well ... is there any detail i need to pay attention to?

**Reply by jackchen8722** [reply]
thanks,I have fixed it to use "uart_poll_out" not "uart_fifo_fill" to send data,and if you want to use this "uart_fifo_fill" function,you should run this in right context
